src/main/java/org/spongepowered/common/mixin/core/entity/merchant/villager/VillagerEntityMixin.java

```diff
@@ -24,121 +24,74 @@
-        return Optional.ofNullable(this.impl$profession);
-    }
-
-    @Override
-    public void bridge$setProfession(final Profession profession) {
-        this.impl$profession = checkNotNull(profession, "VillagerProfession cannot be null!");
-    }
-
-    @Override
-    public void bridge$setCareer(final Career career) {
-        this.setProfession(((SpongeEntityMetadataType) career.getProfession()).type);
-        this.buyingList = null;
-        this.careerId = ((SpongeCareer) career).type + 1;
-        this.careerLevel = 1;
-        this.getRecipes(null);
-    }
+    @Shadow public abstract VillagerData shadow$getVillagerData();
 
     /**
+     * @author i509VCB - Feb 12, 2020 - 1.14.4
+     * @reason Call trade mutators when adding offers. This covers both Wandering Traders and Villagers.
+     *
      * @author gabizou - January 13th, 2016
      * @reason This overwrites the current method using the multi-dimension array with
      * our {@link VillagerRegistry} to handle career levels and registrations
      * for {@link TradeOfferGenerator}s. Note that this takes over entirely
      * whatever vanilla does, but this allows for maximum customization for
      * plugins to handle gracefully.
      */
-    @SuppressWarnings("unchecked")
-    @Overwrite
-    private void populateBuyingList() { // populateBuyingList
-        // Sponge
-        final List<Career> careers = (List<Career>) this.impl$profession.getCareers();
-
-        // EntityVillager.ITradeList[][][] aentityvillager$itradelist = DEFAULT_TRADE_LIST_MAP[this.getProfession()];
-
-        if (this.careerId != 0 && this.careerLevel != 0) {
-            ++this.careerLevel;
+    @Override
+    public void addTrades(MerchantOffers givenMerchantOffers, VillagerTrades.ITrade[] newTrades, int maxNumbers) {
+        Set<Integer> set = Sets.newHashSet();
+        if (newTrades.length > maxNumbers) {
+            while(set.size() < maxNumbers) {
+                set.add(this.rand.nextInt(newTrades.length));
+            }
         } else {
-            // Sponge change aentityvillager$itradelist to use this.profession.getCareers()
-            this.careerId = this.rand.nextInt(careers.size()) + 1;
-            this.careerLevel = 1;
+            for(int i = 0; i < newTrades.length; ++i) {
+                set.add(i);
+            }
         }
 
-        if (this.buyingList == null) {
-            this.buyingList = new MerchantRecipeList();
+        List<MerchantOffer> newOffers = new ArrayList<>(); // Sponge - Create temporary list to store the new offers
+
+        for(Integer integer : set) {
+            VillagerTrades.ITrade villagertrades$itrade = newTrades[integer];
+            MerchantOffer merchantoffer = villagertrades$itrade.getOffer(this, this.rand); // Sponge - cast "this" to Abstract Villager Entity
+            if (merchantoffer != null) {
+                // Sponge - Add to newOffers list instead
+                //givenMerchantOffers.add(merchantoffer);
```

Comment:
Let me think about this change, because it might be possible to pass in a copy of the `MerchantOffers` at the head of the method (by using a `@ModifyVariable` and doing an injection at the end of the method, allowing us to throw a "here's a copy of the changes", or even redirecting the `add()` to add to our own list, and then at the tail, doing the event processing.

Reply:
I did the latter option in this case for simplicity's sake. It's 1 redirect and tail injection now.